test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetConsoleTestCase.cs

```diff
@@ -25,27 +29,34 @@ public void InstallPackageFromPMC()
                 var solutionService = VisualStudio.Get<SolutionService>();
 
                 solutionService.CreateEmptySolution("TestSolution", pathContext.SolutionRoot);
-                var project = solutionService.AddProject(ProjectLanguage.CSharp, ProjectTemplate.ClassLibrary, ProjectTargetFramework.V46, "TestProject");
+                var project = solutionService.AddProject(ProjectLanguage.CSharp, projectTemplate, ProjectTargetFramework.V46, "TestProject");
+                project.Build();
 
                 var packageName = "TestPackage";
                 var packageVersion = "1.0.0";
-                CreatePackageInSource(pathContext.PackageSource, packageName, packageVersion);
+                Utils.CreatePackageInSource(pathContext.PackageSource, packageName, packageVersion);
 
                 var nugetTestService = GetNuGetTestService();
                 Assert.True(nugetTestService.EnsurePackageManagerConsoleIsOpen());
 
                 var nugetConsole = nugetTestService.GetPackageManagerConsole(project.Name);
 
                 Assert.True(nugetConsole.InstallPackageFromPMC(packageName, packageVersion));
+                project.Build();
```

Comment:
I know it's not easy to know when restore completes which is why build step is added. But this doesn't test the real scenario what if there is some issue with install package and some how its not running restore as part of that?

Reply:
yes i understand your concern but if we go that route, no matter what amount of time we wait to wait for restore, there will always be flakiness. Right now, based on my conversation with @emgarten  , build is the only thing that would wait for restore to complete.

It's a non-trivial problem to solve and i would not want to hold off this PR just because of that - look at what that approach got us from powershell - flakiness all the way and the tests lost their value completely.

If you do have a solution for this, i would gladly like to hear, but based on my digging around, this is the best we could do.

Reply:
Here is what this test is actually doing:
1. Running `Install-Package` from PMC, this should block until restore is complete and committed to disk. Waiting for restore is **not** the issue here.
1. Checking the error lists. This feels like it is testing the test, I'm not sure why this would be needed unless the result of the above install command is broken. It also lets in build errors, but maybe that is a good check.
1. `IsPackageInstalled` calls `Get-Package` to check if the package is installed. **This is probably why the test is unstable.**

Instead of checking the assets file for the installed package what I think is going on behind this is:
1. NuGet calls APIs to update the csproj file with the new PackageReference
1. NuGet writes the project.assets.json file to disk
1. The project system nominates the updated project info
1. `Get-Package` reads the updated PackageSpec from the project system to determine if the package is installed

If the 3rd one doesn't happen in time then it would look like the package wasn't installed.

Building forces the project system to nominate. 

We need either:
* Look at the project.assets.json file instead of using `Get-Package`
* Build to force the nomination, or get a new API from the project system to pull nominations instead of push them
* Store our own version of the PackageSpec instead of waiting for the nomination to round trip our action

Reply:
We are doing 2) right now - building to force the nomination

Reply:
Talked to Rohit offline, he'll try to asset through lock file instead of forcing a build and check through `Get-Package`.

I also looked at the asserts for existing NetCore E2E tests (via PMC) and it has two parts:
1. `Assert-NetCorePackageReference` which tries to read Project ItemGroup to check if `PackageReference` item for this specific id and version exists.
2. `Assert-NetCorePackageNotInLockFile` which checks for `project.assets.json` file and check it's library section for the specified id and version.

I think the real reason of net core tests being flack is the first check because we aren't sure when exactly project system will write it to the disk or save in project. So checking through assets file should be good enough here and that's what is in control of NuGet as well.

Reply:
I'm fine with verifying just the assets file, but verifying that the project system actually writes out the PackageReference like the test was doing with Build is a more complete end to end test.

Reply:
I plan on doing both